sys/targets/targets.go

```diff
@@ -612,42 +619,42 @@ func init() {
 			initTarget(target, OS, arch)
 		}
 	}
-	goarch := runtime.GOARCH
+	arch32, arch64 := splitArch(runtime.GOARCH)
 	goos := runtime.GOOS
 	if goos == "android" {
 		goos = Linux
 	}
 	for _, target := range List[TestOS] {
-		if List[goos] != nil {
-			if host := List[goos][goarch]; host != nil {
-				target.CCompiler = host.CCompiler
-				target.CPP = host.CPP
-				if goos == FreeBSD {
-					// For some configurations -no-pie is passed to the compiler,
-					// which is not used by clang.
-					// Ensure clang does not complain about it.
-					target.CFlags = append(target.CFlags, "-Wno-unused-command-line-argument")
-					// When building executor for the test OS, clang needs
-					// to link against the libc++ library.
-					target.CFlags = append(target.CFlags, "-lc++")
-				}
-				// In ESA/390 mode, the CPU is able to address only 31bit of memory but
-				// arithmetic operations are still 32bit
-				// Fix cflags by replacing compiler's -m32 option with -m31
-				if goarch == S390x {
-					for i := range target.CFlags {
-						target.CFlags[i] = strings.Replace(target.CFlags[i], "-m32", "-m31", -1)
-					}
-				}
-			}
-			if target.PtrSize == 4 && goos == FreeBSD && goarch == AMD64 {
-				// A hack to let 32-bit "test" target tests run on FreeBSD:
-				// freebsd/386 requires a non-default DataOffset to avoid
-				// clobbering mappings created by the C runtime. Since that is the
-				// only target with this constraint, just special-case it for now.
-				target.DataOffset = List[goos][I386].DataOffset
+		if List[goos] == nil {
+			continue
+		}
+		arch := arch64
+		if target.PtrSize == 4 {
+			arch = arch32
+		}
+		host := List[goos][arch]
+		if host == nil {
+			continue
+		}
+		target.CCompiler = host.CCompiler
+		target.CPP = host.CPP
+		target.CFlags = append(append([]string{}, host.CFlags...), target.CFlags...)
+		target.CFlags = processMergedFlags(target.CFlags)
+		// In ESA/390 mode, the CPU is able to address only 31bit of memory but
+		// arithmetic operations are still 32bit
+		// Fix cflags by replacing compiler's -m32 option with -m31
+		if arch == S390x {
+			for i := range target.CFlags {
+				target.CFlags[i] = strings.Replace(target.CFlags[i], "-m32", "-m31", -1)
 			}
 		}
+		if target.PtrSize == 4 && goos == FreeBSD && arch == AMD64 {
+			// A hack to let 32-bit "test" target tests run on FreeBSD:
+			// freebsd/386 requires a non-default DataOffset to avoid
+			// clobbering mappings created by the C runtime. Since that is the
+			// only target with this constraint, just special-case it for now.
+			target.DataOffset = List[goos][I386].DataOffset
```

Comment:
Maybe we should take DataOffset from the host as well?

Reply:
Pushed one more commit, now need your approval again.